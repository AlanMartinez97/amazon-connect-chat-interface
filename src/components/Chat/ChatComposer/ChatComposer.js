// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import React, { Component } from "react";
import styled from "styled-components";
import { FormattedMessage } from "react-intl";
import PT from "prop-types";
import { CONTACT_STATUS } from "connect-constants";

const ChatComposerWrapper = styled.div`
  order: 3;
  background: ${props => props.theme.palette.white};
`;
const EditorHintText = styled.div`
  position: absolute;
  color: ${props => props.theme.palette.lightGray};
  font-family: ${props => props.theme.fonts.light};
  z-index: 1;
  padding: ${props => props.theme.spacing.small};
  pointer-events: none;
`;
const EditableContainer = styled.div`
  outline: none;
  user-select: text;
  word-break: break-word;
  border: 0.5px solid ${props => props.theme.palette.lightGray};
  border-left: 0;
  border-right: 0;
  padding: ${props => props.theme.spacing.small};
  max-height: 80px;
  overflow: auto;
  min-height: 39px;
  z-index: 2;
`;

export default class ChatComposer extends Component {
  static propTypes = {
    addMessage: PT.func,
    onTyping: PT.func,
    contactId: PT.string.isRequired,
    contactStatus: PT.string.isRequired,
    onTypingValidityTime: PT.number
  };

  static defaultProps = {
    // Default props
    onTypingValidityTime: 10 * 1000
  };

  state = {
    message: ""
  };

  constructor(props) {
    super(props);
    this.onTypingCalled = false;
  }

  onTyping = event => {
    let self = this;
    if (!this.hasSameContent(event)) {
      if (this.onTypingCalled) {
        return;
      }
      this.props.onTyping(this.props.contactId, event.target.innerText !== '');
      this.onTypingCalled = true;
      setTimeout(() => {
        self.onTypingCalled = false;
      }, self.props.onTypingValidityTime);
    }
  };

  hasSameContent = event => {
    return event.target.innerText === this.state.message;
  };

  onMessageInput = event => {
    this.onTyping(event);
    this.setState({ message: event.target.innerText });
    if (!event.shiftKey && event.which === 13) {
      event.preventDefault();
      this.sendMessage();
      event.target.innerText = "";
      this.setState({ message: "" });
      return false;
    }
  };

  sendMessage = () => {
    if (this.state.message.toString().trim()) {
      this.props.addMessage(this.props.contactId, {
        text: this.state.message
      });
      this.setState({ message: "" });
    }
  };

  onPaste = e => {
    // Stop data actually being pasted into div
    e.stopPropagation();
    e.preventDefault();
    var text = e.clipboardData.getData("text/plain");
    document.execCommand("insertText", false, text);
  };

  render() {
    return (
      <ChatComposerWrapper>
        {(this.props.contactStatus === CONTACT_STATUS.CONNECTED) && (
          <React.Fragment>
            <EditorHintText>
              {!this.state.message && (
                <FormattedMessage
                  id="connect.chat.composer.hintText"
                  defaultMessage="Type a message"
                />
              )}
            </EditorHintText>

            <EditableContainer
              onInput={this.onMessageInput}
              onKeyPress={this.onMessageInput}
              onKeyDown={this.onMessageInput}
              role="textbox"
              aria-label="type a message"
              tabIndex="0"
              contentEditable="true"
              spellCheck="true"
              onPaste={this.onPaste}
            />
          </React.Fragment>
        )}
      </ChatComposerWrapper>
    );
  }
}
