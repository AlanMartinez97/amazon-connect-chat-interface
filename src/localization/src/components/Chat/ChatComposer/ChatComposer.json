[
  {
    "id": "connect.chat.composer.hintText",
    "defaultMessage": "Type a message",
    "file": "src/components/Chat/ChatComposer/ChatComposer.js",
    "start": {
      "line": 113,
      "column": 16
    },
    "end": {
      "line": 116,
      "column": 18
    }
  }
]