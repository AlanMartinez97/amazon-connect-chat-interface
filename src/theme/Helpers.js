// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

import styled from 'styled-components';

const FlexRowContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: ${props => props.center ? "center": "space-between"};
  align-items: ${props => props.center ? "center": "normal"}; 
`

const FlexColumnContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: ${props => props.center ? "center": "space-between"};
  align-items: ${props => props.center ? "center": "normal"}; 
`

export {
  FlexColumnContainer,
  FlexRowContainer
}