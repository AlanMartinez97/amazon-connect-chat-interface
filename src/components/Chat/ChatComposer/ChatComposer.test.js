import React from 'react';
import ReactDOM from 'react-dom';
import ChatComposer from './ChatComposer';
import { ThemeProvider } from '../../../theme';

// Dummy messages
const messages = [
  {
    id: 1,
    message: "Welcome bro",
    from: "agent"
  }, {
    id: 2,
    message: "Hello bro",
    from: "customer"
  },
  {
    id: 3,
    message: "Visit this for more info https://reactjs.org/docs/dom-elements.html",
    from: "agent"
  }
]

describe('<ChatComposer />', () => {
  test("Style should match the snapshot", () => {
    const tree = createTree(<ThemeProvider><ChatComposer messages={messages} /></ThemeProvider>);
    expect(tree).toMatchSnapshot();
  });
});
