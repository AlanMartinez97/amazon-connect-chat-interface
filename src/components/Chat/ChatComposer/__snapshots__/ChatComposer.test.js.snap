// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatComposer /> Style should match the snapshot 1`] = `
.c0 {
  -webkit-order: 3;
  -ms-flex-order: 3;
  order: 3;
  background: #FFFFFF;
}

<div
  className="c0"
/>
`;
